File: src/integrity/fragmentValidator.ts
import type { FragmentDescriptor } from "../contracts/manifest"; import { sha256 } from "./sha256";
export async function validateFragment(input:{fragment:FragmentDescriptor; bytes:Uint8Array}): Promise<boolean>{ if(input.fragment.hashAlgorithm.toUpperCase()!=="SHA-256") return false; return (await sha256(input.bytes))===input.fragment.sha256.toLowerCase(); }

File: crates/pontemesh-sdk-core/src/integrity/fragment_validator.rs
use crate::contracts::FragmentDescriptor;
use crate::errors::PontemeshError;
use crate::integrity::sha256_hex;

pub fn validate_fragment(
    fragment: &FragmentDescriptor,
    bytes: &[u8],
) -> Result<(), PontemeshError> {
    if bytes.len() != fragment.size_bytes {
        return Err(PontemeshError::HashMismatch(format!(
            "fragment {} size mismatch: expected {}, got {}",
            fragment.index,
            fragment.size_bytes,
            bytes.len()
        )));
    }
    let digest = sha256_hex(bytes);
    if !digest.eq_ignore_ascii_case(&fragment.sha256) {
        return Err(PontemeshError::HashMismatch(format!(
            "fragment {} sha256 mismatch",
            fragment.index
        )));
    }
    Ok(())
}
